examples/common/text_data.py

```diff
@@ -84,81 +88,63 @@ def __init__(self,
                          batch_size=batch_size)
         self.tokenizer_name = tokenizer_name
         self.max_seq_len = max_seq_len
-        self.group_method = group_method
 
         # Build tokenizer
         os.environ['TRANSFORMERS_NO_ADVISORY_WARNINGS'] = '1'
+        os.environ['TOKENIZERS_PARALLELISM'] = 'false'
         self.tokenizer = transformers.AutoTokenizer.from_pretrained(
             self.tokenizer_name)
-        if self.tokenizer._pad_token is None:
-            # Some tokenizers (e.g. GPT2 tokenizer) have no padding token which causes bugs
-            self.tokenizer._pad_token = self.tokenizer._eos_token
-        # suppress warnings when using group_method='concat' and no truncation
+
+        # suppress warnings when using longer 'max_seq_len'
         self.tokenizer.model_max_length = int(1e30)
+        self.add_special_tokens = None
 
     # How to tokenize a text sample to a token sample
     def _tokenize(self, text_sample):
-        if self.group_method == 'truncate':
-            truncation = True
-            padding = 'max_length'
-            max_length = self.max_seq_len
-        elif self.group_method == 'concat':
-            truncation = False
-            padding = False
-            max_length = None
-        else:
-            raise ValueError(f"Got unknown group_method='{self.group_method}'.")
+        # The first time we tokenize we need to figure out the special token situation,
+        # as pre-concatenated text will already have special tokens
+        if self.add_special_tokens is None:
+            # if the 2nd token is BOS or 2nd-to-last-is EOS, the special tokens
+            # were added as part of pre-concatenating the text, so don't add them again
+            tokens = self.tokenizer(text_sample['text'],
+                                    truncation=False,
+                                    padding=False,
+                                    add_special_tokens=True)
+            if tokens['input_ids'][1] == self.tokenizer.bos_token_id or tokens[
+                    'input_ids'][-2] == self.tokenizer.eos_token_id:
+                self.add_special_tokens = False
+            else:
+                # Normal (not pre-concatenated) text without special tokens added
+                self.add_special_tokens = True
+
+        if self.tokenizer._pad_token is None:
+            # Some tokenizers (e.g. GPT2 tokenizer) have no padding token which causes bugs
+            raise RuntimeError(
+                'If tokenizing on-the-fly, tokenizer must have a pad_token_id')
```

Comment:
Should we enable users to set a `_pad_token`?
either as a kwarg or provide a member method to our class?

Reply:
Since this class is instantiated from configs, I think we should do it as a kwarg. Sound good?

Reply:
I think we actually don't need to add this. We pass along tokenizer kwargs, so the user can just add `pad_token: <PAD>` to their yaml, and they will end up with a tokenizer with a pad token

Reply:
oh maybe we don't pass along all the kwargs...checking

Reply:
ok we don't pass along the kwargs, but we do still automatically override pad with eos if pad is none. that is in `tokenizer.py`. Not sure if we should get rid of that or not...

Reply:
deferring this to another PR, but planning still before release.